jumanji/environments/routing/mmst/env.py

```diff
@@ -0,0 +1,721 @@
+        - nodes_to_connect: jax array (int) of shape (num_agents, num_nodes_per_agent).
+        - step_count: step counter.
+        - step_limit: the number of steps allowed before an episode terminates.
+        - key: PRNG key for random sample.
+
+    - constants definitions:
+        - Nodes
+            - INVALID_NODE = -1: used to check if an agent selects an invalid node.
+                A node may be invalid if its has no edge with the current node or if it is a
+                utility node already selected by another agent.
+            - UTILITY_NODE = -1: utility node (belongs to no agent).
+            - EMPTY_NODE = -1: used for padding.
+                state.connected_nodes stores the path (all the nodes) visited by an agent. Hence
+                it has size equal to the step limit. We use this constant to initialise this array
+                since 0 represents the first node.
+            - DUMMY_NODE = -10: used for tie-breaking if multiple agents select the same node.
+
+        - Edges
+            - EMPTY_EDGE = -1: used for masking edges array.
+               state.node_edges is the graph's adjacency matrix, but we don't represent it
+               using 0s and 1s, we use the node values instead, i.e `A_ij = j` or `A_ij = -1`.
+               Also edges are masked when utility nodes
+               are selected by an agent to make it unaccessible by other agents.
+
+        - Actions encoding
+            - INVALID_CHOICE = -1
+            - INVALID_TIE_BREAK = -2
+            - INVALID_ALREADY_TRAVERSED = -3
+    """
+
+    def __init__(
+        self,
+        generator_fn: Optional[Generator] = None,
+        reward_fn: Optional[RewardFn] = None,
+        step_limit: int = 100,
+        viewer: Optional[Viewer[State]] = None,
+    ):
+        """Create the `MMST` environment.
+
+        Args:
+            generator_fn: `Generator` whose `__call__` instantiates an environment instance.
+                Implemented options are [`SplitRandomGenerator`].
+            reward_fn: class of type `RewardFn`, whose `__call__` is used as a reward function.
+                Implemented options are [`DefualtRewardFn`]. Defaults to `DefaultRewardFn`.
+            step_limit: the number of steps allowed before an episode terminates. Defaults to 50.
+            viewer: `Viewer` used for rendering. Defaults to `MMSTViewer`
+        """
+
+        self._generator_fn = generator_fn or SplitRandomGenerator(
+            num_nodes=36,
+            num_edges=72,
+            max_degree=5,
+            num_agents=3,
+            num_nodes_per_agent=4,
+            max_step=step_limit,
+        )
+
+        self.num_agents = self._generator_fn.num_agents
+        self.num_nodes = self._generator_fn.num_nodes
+        self.num_nodes_per_agent = self._generator_fn.num_nodes_per_agent
+
+        self._reward_fn = reward_fn or DenseRewardFn(reward_values=(1.0, -0.03, -0.03))
+
+        self._renderer = viewer
```

Comment:
naming this variable `_viewer` or `_env_viewer` would be better than renderer. We're also animating using the viewer